### ISSUE TITLE
error handling: Error leaks with preemptive throws

### ISSUE THREAD

**Issue by psahabu** [chapel-lang/chapel#7331]
### Description

Every `Error` in the Chapel error handling system is a class, but to hide this implementation detail from the user, the compiler inserts deletes for the handled error at the end of every catch block.

``` chapel
try {
  throw new SomeError();
} catch e {
  writeln("handling error " + e.msg);
  // compiler inserts delete e
}
```

This works even in cases when the handled error is thrown from the catch block, as the delete is dead code eliminated in a later pass. However, the following pattern leaks the handled `Error`:

``` chapel
try {
  throw new SomeError();
} catch e {
  writeln("handling error " + e.msg);
  throw new DifferentError();
  // delete e is dead code eliminated
  // e is never deleted
}
```

### Solutions

The current compiler provides no easy solutions because there is no way for `e` to be intelligently deleted, short of inspecting every statement in the `catch` block for a `throw` or a throwing call. This is a motivating case for a stronger error memory management story, be it errors as records instead of classes, or Shared/Owned.

### Affected tests
- `errhandling/psahabu/throw-from-catch`
- `errhandling/psahabu/throw-inner-catch`
- `errhandling/psahabu/unmarked-call-catch`

**Comment by mppf** [comment]
@psahabu - I'm not convinced that there's no easy solution. What if we simply put `delete e` in a DeferStmt at the start of every catch block body?

**Comment by psahabu** [comment]
Closed by #7387.
